internal/api/api.go

```diff
@@ -148,19 +153,50 @@ func NewAPIWithVersion(globalConfig *conf.GlobalConfiguration, db *storage.Conne
 					if !api.config.External.AnonymousUsers.Enabled {
 						return unprocessableEntityError(ErrorCodeAnonymousProviderDisabled, "Anonymous sign-ins are disabled")
 					}
-					if _, err := api.limitHandler(limiter)(w, r); err != nil {
+					if _, err := api.limitHandler(limitAnonymousSignIns)(w, r); err != nil {
 						return err
 					}
 					return api.SignupAnonymously(w, r)
 				}
+
+				// apply ip-based rate limiting on otps
+				if _, err := api.limitHandler(limitSignups)(w, r); err != nil {
+					return err
+				}
+				// apply shared rate limiting on email / phone
+				if _, err := sharedLimiter(w, r); err != nil {
+					return err
+				}
 				return api.Signup(w, r)
 			})
 		})
-		r.With(sharedLimiter).With(api.verifyCaptcha).With(api.requireEmailProvider).Post("/recover", api.Recover)
-		r.With(sharedLimiter).With(api.verifyCaptcha).Post("/resend", api.Resend)
-		r.With(sharedLimiter).With(api.verifyCaptcha).Post("/magiclink", api.MagicLink)
+		r.With(api.limitHandler(
+			// Allow requests at the specified rate per 5 minutes
+			tollbooth.NewLimiter(api.config.RateLimitOtp/(60*5), &limiter.ExpirableOptions{
+				DefaultExpirationTTL: time.Hour,
+			}).SetBurst(30),
+		)).With(sharedLimiter).With(api.verifyCaptcha).With(api.requireEmailProvider).Post("/recover", api.Recover)
 
-		r.With(sharedLimiter).With(api.verifyCaptcha).Post("/otp", api.Otp)
+		r.With(api.limitHandler(
+			// Allow requests at the specified rate per 5 minutes
+			tollbooth.NewLimiter(api.config.RateLimitOtp/(60*5), &limiter.ExpirableOptions{
+				DefaultExpirationTTL: time.Hour,
```

Comment:
Surfacing from discussion, h/t @kangmingtay :

- a single config was used as opposed to one for sign up and one from sign in as we can't tell if the user